falcon/media/multipart.py

```diff
@@ -0,0 +1,402 @@
+                if value is None:
+                    return None
+                self._filename = value
+
+        return self._filename
+
+    @property
+    def secure_filename(self):
+        return misc.secure_filename(self.filename)
+
+    @property
+    def name(self):
+        if self._name is None:
+
+            if self._content_disposition is None:
+                value = self._headers.get(b'content-disposition', b'')
+                self._content_disposition = cgi.parse_header(value.decode())
+
+            _, params = self._content_disposition
+            self._name = params.get('name')
+
+        return self._name
+
+    @property
+    def media(self):
+        if self._media is None:
+            handler = self._parse_options.media_handlers.find_by_media_type(
+                self.content_type, 'text/plain')
+            self._media = handler.deserialize(
+                self.stream, self.content_type, None)
+
+        return self._media
+
+
+class MultipartForm:
+
+    def __init__(self, stream, boundary, content_length, parse_options):
+        # NOTE(vytas): More lenient check whether the provided stream is not
+        #   already an instance of BufferedStream.
+        # This approach makes testing both the Cythonized and pure-Python
+        #   streams easier within the same test/benchmark suite.
+        if not hasattr(stream, 'read_until'):
+            if isinstance(stream, request_helpers.BoundedStream):
+                stream = BufferedStream(stream.stream.read, content_length)
+            else:
+                stream = BufferedStream(stream.read, content_length)
+
+        self._stream = stream
+        self._boundary = boundary
+        # NOTE(vytas): Here self._dash_boundary is not prepended with CRLF
+        #   (yet) for parsing the prologue. The CRLF will be prepended later to
+        #   construct the inter-part delimiter as per the RFC 7578, section 4.1
+        #   (see the note below).
+        self._dash_boundary = b'--' + boundary
+        self._parse_options = parse_options
+
+    def __iter__(self):
+        delimiter = self._dash_boundary
+        stream = self._stream
+        max_headers_size = self._parse_options.max_body_part_headers_size
+        remaining_part_count = self._parse_options.max_body_part_count
+
+        while True:
+            # NOTE(vytas): Either exhaust the unused stream part, or skip
+            #   the prologue.
+            stream.pipe_until(delimiter)
+            stream.read(len(delimiter))
+
+            if not delimiter.startswith(CRLF):
+                # NOTE(vytas): RFC 7578, section 4.1.
+                #   As with other multipart types, the parts are delimited with
+                #   a boundary delimiter, constructed using CRLF, "--", and the
+                #   value of the "boundary" parameter.
+                delimiter = CRLF + delimiter
+
+            separator = stream.read_until(CRLF, 2, MultipartParseError)
```

Comment:
How does this approach of passing `MultipartParseError` compare to raising the same error type each time and letting the caller catch the error and do with it what it will?

Reply:
Umm... that approach would be cleaner and more straightforward, I guess.

IIRC I probably did it this way because the part stream's `read_until()` is exposed to API users, so the generic `MultipartParseError` would sound a bit off. But we could maybe define a new error type for `read_until()`, and re-raise that in the multipart handler?